### PR TITLE
feat: support using a partition table accessor of different type than the multiexponentiation result (PROOF- 870)

### PR DESCRIPTION
**Pull request by rnburn**
# Rationale for this change

Support using an accessor to partition table sums with a different curve form than the multiexponentiation result.

This will allow us to use a more memory efficient curve element.

# What changes are included in this PR?

* Makes minor adjustments to functions and code to support varying curve types.

# Are these changes tested?

Yes

### CODE REVIEW

File: sxt/multiexp/pippenger2/in_memory_partition_table_accessor_utility.h [hunk 1]
```diff
@@ -31,8 +31,12 @@ namespace sxt::mtxpp2 {
 //--------------------------------------------------------------------------------------------------
 // make_in_memory_partition_table_accessor
 //--------------------------------------------------------------------------------------------------
-template <bascrv::element T>
-std::unique_ptr<partition_table_accessor<T>> make_in_memory_partition_table_accessor(
+template <class U, bascrv::element T>
```
Comment (jacobtrombetta) [inline sxt/multiexp/pippenger2/in_memory_partition_table_accessor_utility.h:34]: Is there a reason some of the templates are defined by `template <bascrv::element T, class U>` and others by `template <class U, bascrv::element T>`? I wonder if that will lead to to incorrect ordering of template parameters?

File: sxt/multiexp/pippenger2/in_memory_partition_table_accessor_utility.t.cc [hunk 2]
```diff
@@ -43,7 +43,7 @@ TEST_CASE("we can create a partition table accessor from given generators") {
   memmg::managed_array<E> table(table_dev.size());
 
   SECTION("we can create an accessor from a single generators") {
-    auto accessor = make_in_memory_partition_table_accessor<E>(basct::subspan(generators, 0, 1));
+    auto accessor = make_in_memory_partition_table_accessor<E, E>(basct::subspan(generators, 0, 1));
```
Comment (jacobtrombetta) [inline sxt/multiexp/pippenger2/in_memory_partition_table_accessor_utility.t.cc:46]: If you added a default template parameter, `template <bascrv::element T, class U = T>`, could you avoid needing to use `<E, E>` in cases where the accessor to partition data sums and the MSM result are the same type?

File: sxt/multiexp/pippenger2/in_memory_partition_table_accessor_utility.h (hunk 1 above)
Comment (rnburn) [reply sxt/multiexp/pippenger2/in_memory_partition_table_accessor_utility.h:34]: It's possible to deduce the T type here, but the U type needs to be specified; that's why it's first.

I don't think there's a type mismatch anywhere. In general, type ordering can be different for different functions.

**Review by jacobtrombetta — APPROVED** [review]

File: sxt/multiexp/pippenger2/in_memory_partition_table_accessor_utility.t.cc (hunk 2 above)
Comment (rnburn) [reply sxt/multiexp/pippenger2/in_memory_partition_table_accessor_utility.t.cc]: I made an adjustment that lets allows you to call the function with only a single template argument if the types are the same.

**Pull request merged**